File: stacks/largest_rectangle_in_histogram_best.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        s = []
        res = 0

        for i in range(n):

            while s and heights[s[-1]] >= heights[i]:
                tp = s.pop()

                width = i if not s else i - s[-1] - 1

                res = max(res, heights[tp] * width)

            s.append(i)

        while s:
            tp = s.pop()
            width = n if not s else n - s[-1] - 1
            res = max(res, heights[tp] * width)

        return res

File: stacks/largest_rectangle_in_histogram_best.rb
def largest_rectangle_area(heights)
    n = heights.size
    stack = []
    res = 0

    n.times do |i|
        while !stack.empty? && heights[stack[-1]] >= heights[i]
            top = stack.pop()

            width = stack.empty? ? i : i -r stack.last - 1

            res = [res, heights[top] * width].max
        end

        stack.push(i)
    end

    while !stack.empty?

        top = stack.pop
        width = stack.empty? ? n : n - stack[-1] - 1
        res = [res, heights[top] * width].max

    end
    res

end